### PR TITLE
[NPU][diffusion]: support parallel decoding of qwen-image

### PR DESCRIPTION
**Pull request by gxxx-hum**
## Motivation
Parallel VAE decoding reduces peak memory for Qwen-Image and improves memory stability on large-resolution image generation workloads.

In our comparison against main, resolutions above 1024x1024 have a high chance of running into OOM on the baseline path. With this change, peak memory is reduced by about 5% in our tests, while end-to-end performance remains effectively unchanged.

## Modifications
* Qwen-Image VAE config
  * Enable parallel VAE decoding by default for the Qwen-Image VAE configuration so that autoencoder_kl_qwenimage uses the ParallelTiledVAE path when SP world size > 1.
* Qwen-Image VAE runtime
  * Wire Qwen-Image VAE decoding to the existing parallel_tiled_decode implementation when use_parallel_decode is enabled and sequence-parallel world size is greater than 1.
  * Keep the non-parallel _decode path unchanged and still available as a fallback for environments without SP or when parallel decode is explicitly disabled. 
* Gather decoded tile payloads on rank 0 only, merge once, and broadcast the final decoded output to all ranks.
* Vectorize `blend_v`, `blend_h` to replace Python loops with tensor operations and reduce host-side overhead.

## Accuracy Tests
startup script:
```
nohup sglang serve \
--model-path Qwen-Image-2512 \
--num-gpus 2 \
--sp-degree 2 \
--port 8080 \
--host 0.0.0.0 > sglang.log 2>&1 & 
```

test script:
```
import requests

response=requests.post(
   "http://localhost:8080/v1/images/generations",
   json={
      "prompt": "a black and white cat wearing a princess tiara",
      "size": "1024x1024",
      "num_inference_steps": 50,
      "response_format": "b64_json",
      "seed": 42,
      "n":1
   }
)
```
The current version would have caused an OOM (Out of Memory) problem，the error log:
```
[DecodingStage] Error during execution after 11612.2197 ms: The Inner error is reported as above. The process exits for this inner error, and the current working operator name is Conv2D.
Since the operator is called asynchronously, the stacktrace may be inaccurate. If you want to get the accurate stacktrace, please set the environment variable ASCEND_LAUNCH_BLOCKING=1.
Note: ASCEND_LAUNCH_BLOCKING=1 will force ops to run in synchronous mode, resulting in performance degradation. Please unset ASCEND_LAUNCH_BLOCKING in time after debugging.
[ERROR] 2026-03-16-03:20:06 (PID:97494, Device:0, RankID:0) ERR00100 PTA call acl api failed.
[PID: 97494] 2026-03-16-03:20:04.278.899 Memory_Allocation_Failure(EL0004): Failed to allocate memory requested by GE module.
        Possible Cause: Available memory is insufficient.
        Solution: Close applications not in use.
        TraceBack (most recent call last):
        Call rtMalloc fail, purpose: page caching, type = 2, size:805371904, device_id:0[FUNC:Alloc][FILE:device_allocator.cc][LINE:72]
        Call rtMalloc fail, purpose: page caching, type = 2, size:402718720, device_id:0[FUNC:Alloc][FILE:device_allocator.cc][LINE:72]
        [Exec][Op]Execute op failed. op type = Conv2D, ge result = 4294967295[FUNC:ReportCallError][FILE:acl_op_executor_log_inner.cpp][LINE:124]]
```
after vae parallel decoding adaptation:
image

## Benchmarking and Profiling
Since 1024x1024 images will cause OOM, 768x768 images were chosen for performance testing and comparison.
|  | decoding time | peak gpu memory(GB) | peak allocated memory(GB) | memory pool overhead(GB) | remaining gpu memory at peak(GB)|
|---|---|---|---|---|---|
| main | 0.1101s | 58.39 | 43.22 | 15.17 | 2.57 |
| pr | 0.2196 | 54.57 | 42.62 | 11.95 | 6.39

Parallel VAE decoding cuts peak GPU memory by 6.5% (with remaining headroom up by about 2.5× and memory‑pool overhead down by 21.2%), while increasing decode latency modestly from 0.1101 s to 0.2196 s.

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review Process

1. Ping Merge Oncalls to start the PR flow. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - `/tag-run-ci-label`, `/rerun-failed-ci`, `/tag-and-rerun-ci`
4. After green CI and required approvals, ask Merge Oncalls to merge.

### CODE REVIEW

File: python/sglang/multimodal_gen/runtime/models/vaes/autoencoder_kl_qwenimage.py [hunk 1]
```diff
@@ -1085,19 +1100,15 @@ def tiled_encode(self, x: torch.Tensor) -> AutoencoderKLOutput:
         enc = torch.cat(result_rows, dim=3)[:, :, :, :latent_height, :latent_width]
         return enc
 
-    def tiled_decode(self, z: torch.Tensor, return_dict: bool = True) -> Union[DecoderOutput, torch.Tensor]:
+    def tiled_decode(self, z: torch.Tensor) -> torch.Tensor:
```
Comment (zhuyijie88) [inline python/sglang/multimodal_gen/runtime/models/vaes/autoencoder_kl_qwenimage.py]: why delete return_dict?

Comment (gxxx-hum) [reply python/sglang/multimodal_gen/runtime/models/vaes/autoencoder_kl_qwenimage.py]: Actually, I've always thought that return_dict is useless：
1. the only chain is decode -> _decode -> tiled_decode and the decode already normalize the output to a tensor (by accessing .sample when needed). 
2. the abstract base class ParallelTiledVAE defines `decode` and `tiled_decode` both return tensor.

File: python/sglang/multimodal_gen/runtime/models/vaes/autoencoder_kl_qwenimage.py [hunk 2]
```diff
@@ -956,13 +962,22 @@ def encode(
 
         return posterior
 
-    def _decode(self, z: torch.Tensor, return_dict: bool = True):
+    def _decode_with_parallel_dispatch(self, z: torch.Tensor) -> DecoderOutput:
+        if self.use_parallel_decode and get_sp_world_size() > 1:
+            num_frame = z.shape[2]
+            num_sample_frames = (num_frame - 1) * self.temporal_compression_ratio + 1
+            decoded = ParallelTiledVAE.parallel_tiled_decode(self, z)[:, :, :num_sample_frames]
```
[the remark this answers is not in the document]
Comment (gxxx-hum) [reply python/sglang/multimodal_gen/runtime/models/vaes/autoencoder_kl_qwenimage.py]: Applied, thanks for the suggestions.

File: python/sglang/multimodal_gen/runtime/models/vaes/autoencoder_kl_qwenimage.py [hunk 3]
```diff
@@ -841,6 +844,9 @@ def __init__(
             .to(cuda_device, dtype)
         )
 
+        self.use_parallel_decode = getattr(config, "use_parallel_decode", False)
```
Comment (ping1jing2) [inline python/sglang/multimodal_gen/runtime/models/vaes/autoencoder_kl_qwenimage.py]: please avoid using `getattr/setattr` in SGLang

Comment (gxxx-hum) [reply python/sglang/multimodal_gen/runtime/models/vaes/autoencoder_kl_qwenimage.py]: Thanks for pointing this out. I’ll change it to explicit config access and avoid getattr/setattr .

File: python/sglang/multimodal_gen/configs/models/vaes/qwenimage.py [hunk 4]
```diff
@@ -38,6 +38,8 @@ class QwenImageVAEConfig(VAEConfig):
     use_temporal_tiling: bool = False
     use_parallel_tiling: bool = False
 
+    use_parallel_decode: bool = True
```
Comment (mickqian) [inline python/sglang/multimodal_gen/configs/models/vaes/qwenimage.py]: better make this False by default

File: python/sglang/multimodal_gen/runtime/models/vaes/autoencoder_kl_qwenimage.py [hunk 5]
```diff
@@ -956,30 +965,43 @@ def encode(
 
         return posterior
 
-    def _decode(self, z: torch.Tensor, return_dict: bool = True):
+    def _decode_with_parallel_dispatch(self, z: torch.Tensor) -> DecoderOutput:
+        if self.use_parallel_decode and get_sp_world_size() > 1:
```
Comment (mickqian) [inline python/sglang/multimodal_gen/runtime/models/vaes/autoencoder_kl_qwenimage.py:967]: `get_world_size()` maybe better, if model_parallel is not enabled, we could always use all the gpus we have

File: python/sglang/multimodal_gen/runtime/models/vaes/autoencoder_kl_qwenimage.py [hunk 6]
```diff
@@ -996,29 +1018,207 @@ def decode(self, z: torch.Tensor, return_dict: bool = True) -> Union[DecoderOutp
...
+            torch.arange(blend_extent, device=b.device, dtype=b.dtype) / blend_extent
+        ).view(1, 1, 1, 1, blend_extent)
+        b[:, :, :, :, :blend_extent] = (
+            a[:, :, :, :, -blend_extent:] * (1 - weight)
+            + b[:, :, :, :, :blend_extent] * weight
+        )
         return b
 
+    def parallel_tiled_decode(self, z: torch.FloatTensor) -> torch.FloatTensor:
+        world_size, rank = get_sp_world_size(), get_sp_parallel_rank()
```
Comment (mickqian) [inline python/sglang/multimodal_gen/runtime/models/vaes/autoencoder_kl_qwenimage.py]: ditto

File: python/sglang/multimodal_gen/runtime/models/vaes/autoencoder_kl_qwenimage.py [hunk 7]
```diff
@@ -996,29 +1018,207 @@ def decode(self, z: torch.Tensor, return_dict: bool = True) -> Union[DecoderOutp
...
+        weight = (
+            torch.arange(blend_extent, device=b.device, dtype=b.dtype) / blend_extent
+        ).view(1, 1, 1, 1, blend_extent)
+        b[:, :, :, :, :blend_extent] = (
+            a[:, :, :, :, -blend_extent:] * (1 - weight)
+            + b[:, :, :, :, :blend_extent] * weight
+        )
         return b
 
+    def parallel_tiled_decode(self, z: torch.FloatTensor) -> torch.FloatTensor:
```
Comment (mickqian) [inline python/sglang/multimodal_gen/runtime/models/vaes/autoencoder_kl_qwenimage.py]: looks quite the same with `ParallelTiledVAE.parallel_tiled_decode`, `AutoencoderKLQwenImage` is already a subclass of `ParallelTiledVAE`, why do we need a new method here?

**Review by mickqian — DISMISSED** [review]

File: python/sglang/multimodal_gen/configs/models/vaes/qwenimage.py (hunk 4 above)
Comment (gxxx-hum) [reply python/sglang/multimodal_gen/configs/models/vaes/qwenimage.py]: I originally set it to True for consistency with Wan. But since this is a new path for QwenImage, defaulting to False is safer. I’ll change it to False.

File: python/sglang/multimodal_gen/runtime/models/vaes/autoencoder_kl_qwenimage.py (hunk 5 above)
Comment (gxxx-hum) [reply python/sglang/multimodal_gen/runtime/models/vaes/autoencoder_kl_qwenimage.py:967]: Do you prefer Qwen-Image to use world-group-based parallel decode while Wan stays SP-based, or should we unify both paths under the same group semantics?

Comment (mickqian) [reply python/sglang/multimodal_gen/runtime/models/vaes/autoencoder_kl_qwenimage.py:967]: I think `get_sp_world_size()` is better here

Comment (gxxx-hum) [reply python/sglang/multimodal_gen/runtime/models/vaes/autoencoder_kl_qwenimage.py:967]: Thanks，I‘ll leave the current unchanged

File: python/sglang/multimodal_gen/runtime/models/vaes/autoencoder_kl_qwenimage.py (hunk 7 above)
Comment (gxxx-hum) [reply python/sglang/multimodal_gen/runtime/models/vaes/autoencoder_kl_qwenimage.py]: It is not the same as the base，the base path uses all-gather and merges on every rank, while qwen-Image gathers to rank 0, merges once on rank 0, and then broadcasts the final result.
I scoped it to the Qwen-Image subclass to avoid changing the shared base behavior for other subclasses, especially since Wan may be affected.Also, on NPU I did observe a noticeable peak-memory reduction.
I can first move it back to the base class and test the impact. If it does affect other subclasses, I’ll refactor the shared front half and abstract the collect/merge part to reduce duplicated code.

**Review by ping1jing2 — APPROVED** [review]

File: python/sglang/multimodal_gen/runtime/models/vaes/autoencoder_kl_qwenimage.py (hunk 7 above)
Comment (mickqian) [reply python/sglang/multimodal_gen/runtime/models/vaes/autoencoder_kl_qwenimage.py]: we should make the new approach a method of `ParallelTiledVAE`, providing an alternative option.

How does these two approaches differ, in terms of applicable scenarios, performance?

File: python/sglang/multimodal_gen/runtime/models/dits/qwen_image.py [hunk 8]
```diff
@@ -826,30 +829,52 @@ def _modulate(
...
-            if is_scale_residual:
-                if not residual_x.is_contiguous():
-                    residual_x = residual_x.contiguous()
-                if not gate_x.is_contiguous():
-                    gate_x = gate_x.contiguous()
-                x, residual_out, gate_result = (
-                    fuse_residual_layernorm_scale_shift_gate_select01_kernel(
+            # ROCm currently fails to compile the select01 Triton kernel, so
+            # keep using the torch.where fallback there.
+            if x.is_cuda and not current_platform.is_hip():
```
Comment (mickqian) [inline python/sglang/multimodal_gen/runtime/models/dits/qwen_image.py:834]: `if x.is_cuda and not current_platform.is_hip():` this pattern is appearing repeatedly recently, please move the implementation to `platform.py`, including all the changes recently

File: python/sglang/multimodal_gen/runtime/models/vaes/autoencoder_kl_qwenimage.py (hunk 7 above)
Comment (gxxx-hum) [reply python/sglang/multimodal_gen/runtime/models/vaes/autoencoder_kl_qwenimage.py]: I see, thanks. You mean this should be a selectable strategy in `ParallelTiledVAE`, rather than a Qwen-Image-specific override. I can further make it a base-level alternative option while keeping the default behavior unchanged.

This new path seems to be a better fit for image-generation scenarios. I’ll also run more comparative experiments to better characterize the differences and refine this part.

File: python/sglang/multimodal_gen/runtime/models/dits/qwen_image.py (hunk 8 above)
Comment (gxxx-hum) [reply python/sglang/multimodal_gen/runtime/models/dits/qwen_image.py:834]: I’ll move it under `runtime/platforms` and clean up the recent similar cases there as well.
